### PR TITLE
feat(Collector): add `ignore` event

### PR DESCRIPTION
**Pull request by ImRodry**
**Please describe the changes this PR makes and why it should be merged:**

This PR adds an option to the collector class with a function to be called when an item is filtered out of the collector, useful when working with interactions and sending error messages for unwanted ones.

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

**Comment by almeidx** [comment]
Shouldn't this be an event instead?

**Review by Jiralite — CHANGES_REQUESTED** [review]
"Discarded" would be a better word choice here. It would make for a more concise description and wouldn't leave `onFiltered` open to interpretation.

**Comment by ImRodry** [comment]
I disagree. I don't see how `onFiltered` is open for interpretation and it makes sense seeing as the other option is called filter

**Comment by ImRodry** [comment]
> Shouldn't this be an event instead?

Personally I would prefer it to be like this but if the maintainers prefer an event, I can also do that

**Comment by Jiralite** [comment]
> Shouldn't this be an event instead?

I believe this should be an event instead.

I'm also not sure about this idea altogether... the purpose of a collector is to collect items. In here, we're specifically looking at what we're discarding. Seems one should handle this in the filter instead?

**Comment by ImRodry** [comment]
The way I see it is that the filter is only meant to return a boolean and not do anything with the received data. You can technically put all your code in the filter which would then make you question whether a collect event is needed at all (or a filter, depending from the perspective). I feel like having this option would make people’s code more organized as you’re telling the collector what to do when it removes an item from the get-go

**Comment by monbrey** [comment]
+1 for this concept but also +1 for it being an event

**Comment by ImRodry** [comment]
> +1 for this concept but also +1 for it being an event

what should the event be called? `filtered`?

**Comment by almeidx** [comment]
To me at least, "filtered" makes it sound like it would only trigger when things pass the filter (as in are accepted through the filter), as per the definition of the word:
> 1. pass (a liquid, gas, light, or sound) through a device to remove unwanted material.
> "the eye filters out ultraviolet radiation"

I believe @Jiralite's suggestion makes it more clear

**Comment by BaumianerNiklas** [comment]
> what should the event be called? `filtered`?

Regardless of which word gets chosen here, shouldn't it use the infinitive as that's used in every other event name? So `filter` instead of `filtered` (if that's what gets decided on), like how it's `collect` instead of `collected`.

**Comment by ImRodry** [comment]
> > what should the event be called? `filtered`?
> 
> Regardless of which word gets chosen here, shouldn't it use the infinitive as that's used in every other event name? So `filter` instead of `filtered` (if that's what gets decided on), like how it's `collect` instead of `collected`.

I understand the thought process but I don't think we should use that same logic here. We use `collect` for when the collector **collects** something. Here, using the name `filter` for the event could make it seem like it's called when the filter is called, which is not the point. I'll stick to `filtered` for now as this implies the event is called when something is filtered, and not when anything reaches the filter. Will also make it an event in the next commit

**Comment by BaumianerNiklas** [comment]
That's reasonable, though still a huge inconsistency I feel like. I would actually agree with going with something like `discard` as an event name then

**Comment by ImRodry** [comment]
I just don't feel like discard has any correlation with the options on the collector. A filter filters items in and out. Items filtered in are collected and sent through the collect event. Items filtered out are filtered and sent through the filtered event. This is also in line with the meaning of the filtered word when used as an adjective: https://en.wiktionary.org/wiki/filtered#Adjective

**Comment by monbrey** [comment]
The event names are not adjectives. The fact you had to add the additional qualifiers "filtered **in**" and "filtered **out**" demonstrates exactly why "filtered" is not clear enough on its own.

The filter function makes a choice to collect, or discard/ignore/whatever the item. However every single received item must be filtered in order for it to make that decision.

**Comment by almostSouji** [comment]
"filtered" does not fit the general event naming conventions, that would be on "filter".
However, "filter" is also ambiguous in meaning, is the element kept or ignored? Is it filtered for or filtered out?

Discard isn't the right choice either.
It's semantically very hard to discern from "dispose", which also already exists as an event and option.
Discard as well as dispose has a strong connotation to trash, "throwing away" or "getting rid of something" (that was present before). As such it fits the "dispose" event, which removes something from the collected items.

Here, however, we need something to express the opposite of "collect". So what does this actually represent, what happens here.
Well, a collector filters incoming events and either "collect"s the structure or does not collect it. Trying to find the opposite of collecting, the best I can come up with is "ignore". The structure is ignored, not collected.

This clearly sets it apart from the semantics of throwing things away (dispose is no longer confusable with discard) and clearly implies which end of the filter something comes out of. If something is collected it's kept, if something is ignored, it's not collected.

**Comment by almostSouji** [comment]
Separating this part out, because agreement above does not need to reflect this side note as well:
Overall my opinion on an event with this functionality: 👍 
While it can arguably be done with a filter function with side effects right now, I'm personally not a huge fan of side effects, when they can be avoided. Tends to be a pain when debugging and trying to evaluate correctness.

**Comment by ImRodry** [comment]
Applied @almostSouji's suggestion for the name of the event

File: packages/discord.js/src/structures/interfaces/Collector.js
Comment (imranbarbhuiya) [inline packages/discord.js/src/structures/interfaces/Collector.js:65]: are there any reason why r u storing it in a collection?
bc I don't see any use case here like `collected` collection have which is used in `end` event

File: packages/discord.js/typings/index.d.ts
Comment (imranbarbhuiya) [inline packages/discord.js/typings/index.d.ts:805]:
```suggestion
  public ignored: Collection<K, V>;
```

**Review by imranbarbhuiya — CHANGES_REQUESTED** [review]

File: packages/discord.js/typings/index.d.ts
Comment (ImRodry) [reply packages/discord.js/typings/index.d.ts:805]: god damn it

File: packages/discord.js/src/structures/interfaces/Collector.js
Comment (ImRodry) [reply packages/discord.js/src/structures/interfaces/Collector.js:65]: For the exact same reason there is a collected collection. It's useless to me, but if someone added it there must be a use, and you may always want to do something with the ignored items at the end, who knows. Either way, the possibility is there

File: packages/discord.js/src/structures/interfaces/Collector.js
Comment (muchnameless) [inline packages/discord.js/src/structures/interfaces/Collector.js:108]:
```suggestion

    if (collectedId) {
        const filterResult = await this.filter(...args, this.collected);
```
the filter should only be called for items that pass the collect check

**Review by muchnameless — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/structures/interfaces/Collector.js
Comment (imranbarbhuiya) [reply packages/discord.js/src/structures/interfaces/Collector.js:65]: Well, then I'll prefer adding it in the end event then. `collected, ignored, reason`.

Comment (kyranet) [reply packages/discord.js/src/structures/interfaces/Collector.js:65]: JFYI, I'm not approving unless this is removed.

This collection is useless, it can and **will** result on extremely large memory usage, for something not a single soul cares about. Entries that are ignored should stay ignored, no mirroring from the collected items, which are something we care about.

Comment (ImRodry) [reply packages/discord.js/src/structures/interfaces/Collector.js:65]: You're right, I didn't think of that

**Comment by Syjalo** [comment]
[REQUEST] Add `ignored` collection and add it to the last parameter of `filter` function.
Maybe also provide `collected` and `ignored` collections in `collect` and `ignore` events.

**Comment by ImRodry** [comment]
I had an `ignored` collection in the first commit of this PR but was asked to remove it because it would be taking up memory unnecessarily, especially when receiving a large number of items in a collector. As for the other change I won't comment on that because that would be out of the scope of this PR

**Review by kyranet — APPROVED** [review]

**Comment by BenPatersonxx** [comment]
@iCrawl @SpaceEEC please review

**Review by SpaceEEC — APPROVED** [review]

**Review by iCrawl — APPROVED** [review]

**Pull request merged**
